bookops_worldcat/_session.py

```diff
@@ -35,6 +36,12 @@ def __init__(
         """
         super().__init__()
 
+        # allow session to retry a request up to 3 times
+        retries = Retry(
+            total=3, backoff_factor=0.5, status_forcelist=[406, 429, 500, 502, 503, 504]
+        )
+        self.mount("https://", requests.adapters.HTTPAdapter(max_retries=retries))
```

Comment:
@charlottekostelic please take a look at `requests.adapters.HTTPAdapter` code [here](https://github.com/psf/requests/blob/eea3bbf9ac635f465ee6c9903dc57c677952dafd/src/requests/adapters.py#L135C1-L145C59). If I'm reading it correctly `requests` dumbs down what is available in urllib3's `Retry` and only one argument is considered: `max_retries`. `backoff_factor` & `status_foreclist` will be ignored.
See [here](https://github.com/psf/requests/blob/eea3bbf9ac635f465ee6c9903dc57c677952dafd/src/requests/adapters.py#L121-L124) for an example how to mount the adapter and retries.

Reply:
Since it appears there is no control what HTTP error codes trigger a retry, we should avoid blindly setting any retries. Let's leave it to users: `WorldcatSession` should take `max_retries` argument when instantiating with a default value 0.

Reply:
```python
class WorldcatSession(requests.Session):
    def __init__(self, authorization, agent, timeout, max_retries):
        ....
```
It would be nice to put retries logic below timeouts and just before `self._update_authorization()`.
The argument `max_retries` should be then added to `MetadataSession`, and passed to `WorldcatSession`. Something like this:
```python
super().__init__(authorization, agent=agent, timeout=timeout, max_retries=max_retries`
```
`max_retries` in `MetadataSession` should also have default value 0.

Reply:
> @charlottekostelic apologies for not catching this earlier. Please take a look at `requests.adapters.HTTPAdapter` code [here](https://github.com/psf/requests/blob/eea3bbf9ac635f465ee6c9903dc57c677952dafd/src/requests/adapters.py#L135C1-L145C59). If I'm reading it correctly `requests` dumbs down what is available in urllib3's `Retry` and only one argument is considered: `max_retries`. `backoff_factor` & `status_forcelist` will be ignored. See [here](https://github.com/psf/requests/blob/eea3bbf9ac635f465ee6c9903dc57c677952dafd/src/requests/adapters.py#L121-L124) for an example how to mount the adapter and retries.

The Requests documentation has an example that passes additional arguments using a `Retry` object from `urllib3` [here](https://requests.readthedocs.io/en/latest/user/advanced/#example-automatic-retries). My understanding of the documentation is that Requests only accepts one argument (`max_retries`) out of the box but with it is possible to customize retry behavior by passing `max_retries` a `Retry` object. 

I just added a couple of tests to `test_session.py` that I think confirm this.

I like the idea of letting users decide whether to use automatic retries. It could have a default of 0 retries but allow users to customize the number of retries and behavior.

Reply:
Ha! Can you take a look at the `requests` code and figure out if that really is implemented? I suspect I missed things when `.mount()` is called?

Reply:
The [doc string](https://github.com/psf/requests/blob/eea3bbf9ac635f465ee6c9903dc57c677952dafd/src/requests/adapters.py#L110-L116) of the `max_retries` parameter says to import `Retry` from `urllib3` and pass that if you need more granular control over retry behavior. 

In the `requests` [code](https://github.com/psf/requests/blob/eea3bbf9ac635f465ee6c9903dc57c677952dafd/src/requests/adapters.py#L135C1-L145C59), `max_retries` is a `Retry` object created using the [`from_int()`](https://github.com/urllib3/urllib3/blob/main/src/urllib3/util/retry.py#L266-L284) class method. In `urllib3`, `from_int()` is used to support backwards compatibility of older retry formats.

Reply:
Thanks! Looks like everything was cleared up and your approach should stay.